dbms/src/Flash/Pipeline/Schedule/TaskThreadPoolMetrics.h

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/TiFlashMetrics.h>
+
+#include <atomic>
+
+namespace DB
+{
+// TODO support more metrics after profile info of task has supported.
+template <bool is_cpu>
+class TaskThreadPoolMetrics
+{
+public:
+    TaskThreadPoolMetrics()
+    {
+        if constexpr (is_cpu)
+        {
+            GET_METRIC(tiflash_pipeline_scheduler, type_cpu_pending_tasks_count).Set(0);
+            GET_METRIC(tiflash_pipeline_scheduler, type_cpu_executing_tasks_count).Set(0);
+            GET_METRIC(tiflash_pipeline_scheduler, type_cpu_task_thread_pool_size).Set(0);
+        }
+        else
+        {
+            GET_METRIC(tiflash_pipeline_scheduler, type_io_pending_tasks_count).Set(0);
+            GET_METRIC(tiflash_pipeline_scheduler, type_io_executing_tasks_count).Set(0);
+            GET_METRIC(tiflash_pipeline_scheduler, type_io_task_thread_pool_size).Set(0);
+        }
+        setMaxTimeMetrics(0);
+    }
+
+    void incPendingTask(size_t task_count)
+    {
+        if constexpr (is_cpu)
+        {
+            GET_METRIC(tiflash_pipeline_scheduler, type_cpu_pending_tasks_count).Increment(task_count);
+        }
+        else
+        {
+            GET_METRIC(tiflash_pipeline_scheduler, type_io_pending_tasks_count).Increment(task_count);
+        }
+    }
+
+    void decPendingTask()
+    {
+        if constexpr (is_cpu)
+        {
+            GET_METRIC(tiflash_pipeline_scheduler, type_cpu_pending_tasks_count).Decrement();
+        }
+        else
+        {
+            GET_METRIC(tiflash_pipeline_scheduler, type_io_pending_tasks_count).Decrement();
```

Comment:
We can define a macro to avoid this bunch of code

Reply:
good point, updated.